staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go

```diff
@@ -111,54 +113,30 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 		return nil, nil, cmdutil.AddSourceToErr(fmt.Sprintf("retrieving original configuration from:\n%v\nfor:", obj), source, err)
 	}
 
-	var patchType types.PatchType
 	var patch []byte
-	var lookupPatchMeta strategicpatch.LookupPatchMeta
-	var schema oapi.Schema
-	createPatchErrFormat := "creating patch with:\noriginal:\n%s\nmodified:\n%s\ncurrent:\n%s\nfor:"
-
-	// Create the versioned struct from the type defined in the restmapping
-	// (which is the API version we'll be submitting the patch to)
-	versionedObject, err := scheme.Scheme.New(p.Mapping.GroupVersionKind)
-	switch {
-	case runtime.IsNotRegisteredError(err):
-		// fall back to generic JSON merge patch
-		patchType = types.MergePatchType
-		preconditions := []mergepatch.PreconditionFunc{mergepatch.RequireKeyUnchanged("apiVersion"),
-			mergepatch.RequireKeyUnchanged("kind"), mergepatch.RequireMetadataKeyUnchanged("name")}
-		patch, err = jsonmergepatch.CreateThreeWayJSONMergePatch(original, modified, current, preconditions...)
+
+	patchType, err := p.getPatchType(p.Mapping.GroupVersionKind)
+	if err != nil {
+		return nil, nil, cmdutil.AddSourceToErr(fmt.Sprintf("getting patch type for %v:", p.Mapping.GroupVersionKind), source, err)
```

Comment:
`source` is given to the `patchSimple` method merely so that we can build the error, which we almost always build in the same way, I think we can do better than that. The `AddSourceToErr` function isn't even really used properly.

I think you can:
1. Stop passing "source" to `patchSimple`, change the calling function to call `AddSourceToErr` with the returned error.
2. Only return mere, normal errors from this method
3. Don't pass `source` anymore to `buildMergePatch`

I suspect you could always use the `createPatchErrFormat` in the calling function, without having to hard-code it at the top of the file.

Reply:
I removed `source` field from `patchSimple` because you are right. We are just passing it to pass `AddSourceToErr`.

Only downside of this change is we need to discriminate errors(ie. https://github.com/kubernetes/kubernetes/blob/609db7ed0b1f2839e414c17d29fe4d76edc994bd/staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go#L105 and https://github.com/kubernetes/kubernetes/blob/609db7ed0b1f2839e414c17d29fe4d76edc994bd/staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go#L111) for backwards compatibility. Thus, I started passing error verb to caller function.

Reply:
"backwards compatibility", how bad is it if we change the error a little bit? Returning the two values is pretty ugly :-/
I don't think we make any guarantees that the error strings have to be constant (it may break some tests possibly, but I'm curious).

Reply:
Maybe I can just return err(after wrapping them with errVerb using errors.Wrap...) to caller. That will change error messages but this will be much more clearer.

Reply:
I think that's fine, you can ask sig-cli @seans3 @soltysh